in_toto/verifylib.py

```diff
@@ -1026,8 +1025,9 @@ def verify_disallow_rule(rule, source_artifacts_queue):
       source_artifacts_queue, rule_data["pattern"])
 
   if len(matched_artifacts):
-    raise RuleVerificationError("Rule '{0}' failed, pattern matched disallowed"
-        " artifacts: '{1}' ".format(" ".join(rule), matched_artifacts))
+    raise RuleVerificationError("Rule '{0}' failed, rule pattern matches the"
+        " following artifacts of the artifact queue, which is disallowed:"
+        " '{1}' ".format(" ".join(rule), matched_artifacts))
```

Comment:
TBH, I don't think this changes a lot. It does rephrase the error message slightly, but does not add any information content as requested in #228.

Since `len(matched_artifacts)`, which decides whether `RuleVerificationError` is raised or not, heavily depends on earlier rule processing, it might be helpful to cache the results of the earlier rule processing and add them to the error message. What do you think? (cc'ing @trishankatdatadog)

Reply:
I think that would be a good idea. How would we fetch the results of the earlier rule processing? Should we just make a call to one of the `verify_*_rule` functions? If so, which one is it?

Reply:
The calls to `verify_*_rule` are already happening. You'd have to cache their results and present it to the user in case an error occurs. You can either use a global variable which you reset in each run of `verify_*_rule`. Or you keep such a variable local in `verify_item_rules`, append the results from the calls to the `verify_*_rule` functions and always pass it to `verify_delete_rule` so that it can be displayed there in case of an error.

In general I am not a fan of using global variables, but it might be okay in this case. Use whatever seems cleaner to you.

Reply:
Okay I think I understand the gist of what you are saying. I have just committed more changes that should hopefully implement the caching of earlier rule processing results. I am a bit unsure about whether I got the right idea from your comments so please let me know if you this is not what you had in mind.

Reply:
Yes, this looks good pretty much like what I had in mind. However, you should not add extra calls to `verify_disallow_rule`, this completely changes the rule processing logic, which we don't want.

Also, just adding the updated queues to your traceback list might not be informative enough.
Please try to add some more context. You don't have to code it right away, maybe you can start by jotting down a data structure here (can be pseudo-code) with all the info you consider relevant for someone you receives the error.

Reply:
Okay I get your point. Just to clarify, what do the updated queues represent? Do they contain all the artifact patterns that were successfully or were failed to be processed? Would making the traceback list into a traceback dictionary (with the keys being the rules as they are called and the values as the updated queues) provide enough context? Or should I add even more info?

Reply:
I like your thinking, adding the rules is definitely a good idea. I would, however, keep the list as base structure to preserve the order. But it could be a list of lists or dictionaries. Please take a closer look at the various `verify_*_rule` functions and the dispatching `verify_item_rules` function to really understand what they are doing and what information is relevant and should therefor be propagated.